tools/codegen/model.py

```diff
@@ -341,11 +335,34 @@ def from_yaml(ei: Dict[str, object], loc: 'Location') -> 'NativeFunction':
             "strictly subsumes the other.  If you wanted to provide an explicit autograd " \
             "implementation, specify CompositeExplicitAutograd; otherwise specify CompositeImplicitAutograd only"
 
+        if structured_delegate:
+            # Structured functions MUST have a dispatch table
+            is_abstract = True
+        else:
+            is_abstract = dispatch.keys() != {DispatchKey.CompositeImplicitAutograd}
+
+        has_composite_implicit_autograd_kernel = DispatchKey.CompositeImplicitAutograd in dispatch.keys()
+        has_composite_explicit_autograd_kernel = DispatchKey.CompositeExplicitAutograd in dispatch.keys()
+
+        # BackendMetadata is used to store any information about a NativeFunction that is backend dependent.
+        # The most obvious information is the kernel name, which usually contains the name of the backend in it for cpu/cuda.
+        # Why is 'structured' included? External backends (e.g. XLA) opt into which ops are structured
+        # independently of which in-tree ops are structured
+        backend_metadata = {k: {func.name: BackendMetadata(
+            kernel=v, structured=structured and is_structured_dispatch_key(k), external=False)} for k, v in dispatch.items()}
+
         # don't care if it exists or not; make it easier to use this function
         # with other yaml parsers that aren't setting __line__ in the dict
         e.pop('__line__', None)
         assert not e, f"leftover entries: {e}"
 
+        # Asserts that we can't do in post_init, because they rely on backend-specific info
```

Comment:
Instead of moving this assert out of post_init, I could have added some properties on NativeFunction to ensure that post_init has enough information to perform the assert.

I couldn't think of a way to do it that preserves the error message without just saving the list of dispatch keys, so I moved the assert here instead.